[InstCombine] Canonicalize zext+overflow check to overflow check if zext's only purpose is to check overflow

AZero13 opened the pull request:
Change processUMulZExtIdiom to also support adds, since the idiom is the same, except with add instead of mul.

Alive2: https://alive2.llvm.org/ce/z/SsB4AK

AZero13 commented:
@topperc What are your thoughts on this?

AZero13 commented:
@RKSimon @dtcxzyw Thoughts?

AZero13 commented:
@topperc What are your thoughts on this since you wrote this originally?

AZero13 commented:
@dtcxzyw Thoughts on this?

topperc commented:
Do you have any real word examples of this pattern? It's a bit of an unusual way to check for unsigned add overflow. The more common way is to check if the result is less than one of the inputs.

AZero13 commented:
> > Do you have any real word examples of this pattern? It's a bit of an unusual way to check for unsigned add overflow. The more common way is to check if the result is less than one of the inputs.
> 
> Well, yes.
> 
> Casting up to a higher value and checking is one way I was taught in programming class.
> 
> AND, not to mention, the work is already here, AND I saw this pattern in systemd's source code iirc.

Also @dtcxzyw Can you run this in opt-benchmark?

AZero13 commented:
@topperc since you are here:

m_NUWAdd is not needed in theory, because this will always be nuw. alive2 works without the nuw tag.

Should I just change it to m_Add, or keep NUWAdd, because any condition that will be tranformed will ALWAYS be nuwadd, well, assuming the rest of the llvm code works properly.

AZero13 commented:
Note that uadd_with_zext_neg_use doesn't have nuw, but it works because it is always given nuw.

topperc commented:
> Note that uadd_with_zext_neg_use doesn't have nuw, but it works because it is always given nuw.



> @topperc since you are here:
> 
> m_NUWAdd is not needed in theory, because this will always be nuw. alive2 works without the nuw tag.
> 
> Should I just change it to m_Add, or keep NUWAdd, because any condition that will be tranformed will ALWAYS be nuwadd, well, assuming the rest of the llvm code works properly.

The NUW is definitely needed for multiplication because there is no check in the code that there are enough bits to hold the full product. The code that sets the nuw flag is responsible for that. For add, we only need one extra bit and a zext has to increase the size by at least 1 bit so the NUWAdd probably isn't explicitly needed, but it should always be there since the add will be visited by InstCombine before the icmp.

AZero13 commented:
> > Note that uadd_with_zext_neg_use doesn't have nuw, but it works because it is always given nuw.
> 
> > @topperc since you are here:
> > m_NUWAdd is not needed in theory, because this will always be nuw. alive2 works without the nuw tag.
> > Should I just change it to m_Add, or keep NUWAdd, because any condition that will be tranformed will ALWAYS be nuwadd, well, assuming the rest of the llvm code works properly.
> 
> The NUW is definitely needed for multiplication because there is no check in the code that there are enough bits to hold the full product. The code that sets the nuw flag is responsible for that. For add, we only need one extra bit and a zext has to increase the size by at least 1 bit so the NUWAdd probably isn't explicitly needed, but it should always be there since the add will be visited by InstCombine before the icmp.

Alright. Thank you!

AZero13 commented:
@topperc

It is in fact used.

AZero13 commented:
@topperc @arsenm Thoughts on this?

arsenm commented:
I thought instcombine deliberately avoided introducing these intrinsics and left it for codegenprepare. CodeGenPrepare seems to not catch this pattern though

AZero13 commented:
Codegenprepare my bad

Also instcombine catches all these patterns I've never heard of codegenprepare doing this

AZero13 commented:
All the patterns for this are caught by instcombine

Not sure where you got the idea codegenprepare did this

AZero13 commented:
> > All the patterns for this are caught by instcombine
> > Not sure where you got the idea codegenprepare did this
> 
> There's a function in CodeGenPrepare that does a different version of an add with overflow pattern `CodeGenPrepare::combineToUAddWithOverflow`

In any case, it is way less burdensome to extend what is there to catched unsigned multiplicaiton overflows.

Moving this functionality to codegenprepare should not just be for add and something I will look into and attempt separately because it seems very involved.

AZero13 commented:
> > All the patterns for this are caught by instcombine
> > Not sure where you got the idea codegenprepare did this
> 
> There's a function in CodeGenPrepare that does a different version of an add with overflow pattern `CodeGenPrepare::combineToUAddWithOverflow`

Anyway, I would say it is better for now to catch these patterns here, rather than codegenprepare.

topperc commented on llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:
```diff
@@ -6622,38 +6626,42 @@ static Instruction *processUMulZExtIdiom(ICmpInst &I, Value *MulVal,
-          TI->setOperand(0, Mul);
+          TI->setOperand(0, Extract);
       } else if (BinaryOperator *BO = dyn_cast<BinaryOperator>(U)) {
         assert(BO->getOpcode() == Instruction::And);
```
This is code tested for Add?

topperc replied:
I don't see any And in your test? Is there some other optimization that creates an And from your test?

arsenm commented:
> Anyway, I would say it is better for now to catch these patterns here, rather than codegenprepare. @topperc what do you think?

It might be easier to add these cases, but it increase the number of places that now need to worry about the add overflow intrinsics. It would probably be better to rewrite this case into the add-with-overflow pattern that CodeGenPrepare does recognize, rather than writing directly to the intrinsics

topperc commented:
> > Anyway, I would say it is better for now to catch these patterns here, rather than codegenprepare. @topperc what do you think?
> 
> It might be easier to add these cases, but it increase the number of places that now need to worry about the add overflow intrinsics. It would probably be better to rewrite this case into the add-with-overflow pattern that CodeGenPrepare does recognize, rather than writing directly to the intrinsics

Does it even need to be CodeGenPrepare? DAGCombine could do this. It's a single basic block pattern. I think the one in CodeGenPrepare is only there because it's cross basic block.

topperc commented:
I have asked you repeatedly in the past to not force push unless you absolutely have to. If you need to update to latest main you should make a merge commit. This is documented in our development policies https://llvm.org/docs/GitHub.html#updating-pull-requests

topperc commented on llvm/test/Transforms/InstCombine/uadd-with-overflow.ll:
```diff
@@ -147,3 +147,76 @@ define { <2 x i32>, <2 x i1> } @fold_simple_splat_constant_with_or_fail(<2 x i32
+  %conv = zext i32 %x to i64
+  %conv1 = zext i32 %y to i64
+  %add = add i64 %conv, %conv1
+  %and = and i64 %add, 65536
```
This is a bit of an odd choice of constant. 65536 has a single bit set. Something closer to 4294967295 would match the intent of the check in the code.

AZero13 commented:
> I have asked you repeatedly in the past to not force push unless you absolutely have to. If you need to update to latest main you should make a merge commit. This is documented in our development policies https://llvm.org/docs/GitHub.html#updating-pull-requests

oh no i was not updating to main. I was adding the and test case to pre-commit and then changed after

AZero13 commented:
> Did you just force push main again? I clicked the "Compare" button on github and it is showing additional changes from main. It makes it very hard to see whether the content of your patch changed. So I have to assume it did and must re-review it all over again.



> Did you just force push main again? I clicked the "Compare" button on github and it is showing additional changes from main. It makes it very hard to see whether the content of your patch changed. So I have to assume it did and must re-review it all over again.

I get it now. Ok. No more force-pushing. I now see the issue.

AZero13 commented:
@topperc ping

topperc reviewed (approved):
LGTM

nikic reviewed (commented):
As @arsenm said, uadd.with.overflow is non-canonical and should not be produced by InstCombine. It should produce a canonical overflow check instead.

AZero13 commented:
Thank you: I don't have merge perms by the way. Can you merge please?

topperc reviewed (changes requested):
Removing my approval per @nikic feedback.

topperc commented:
> What is the canonical overflow check?

Checking if the result of the add is less than one of the inputs.

topperc commented:
Again you force pushed from main after asking for review. Stop doing that.

AZero13 commented:
should I do return IC.replaceInstUsesWith(I, OverflowCheck), or is returning it enough @topperc

AZero13 commented:
@nikic If I need to freeze X, can I just output the intrinsic I don't know.

AZero13 commented:
@topperc ping

AZero13 commented:
@nikic Ping

AZero13 commented:
@topperc ping

arsenm commented on llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:
```diff
@@ -6533,72 +6533,77 @@ bool InstCombinerImpl::OptimizeOverflowCheck(Instruction::BinaryOps BinaryOp,
+                                      const APInt *OtherVal,
+                                      InstCombinerImpl &IC) {
   // Don't bother doing this transformation for pointers, don't do it for
   // vectors.
-  if (!isa<IntegerType>(MulVal->getType()))
+  if (!isa<IntegerType>(Val->getType()))
```
I don't understand why this would skip vectors. Also the comment about pointers doesn't make sense, a pointer type couldn't reach here in the first place

AZero13 replied:
I didn't write this code I just renamed the variable. 

Do you want me to remove it?

arsenm commented on llvm/test/Transforms/InstCombine/saturating-add-sub.ll:
```diff
@@ -2350,4 +2350,57 @@ define i8 @fold_add_umax_to_usub_multiuse(i8 %a) {
   ret i8 %sel
 }
 
+define i32 @uadd_with_zext(i32 %x, i32 %y) {
+; CHECK-LABEL: @uadd_with_zext(
+; CHECK-NEXT:    [[COND:%.*]] = call i32 @llvm.uadd.sat.i32(i32 [[X:%.*]], i32 [[Y:%.*]])
+; CHECK-NEXT:    ret i32 [[COND]]
+;
```
Producing uadd.sat doesn't match the description

AZero13 replied:
Well, that's how instcombine is folding it.

What I do only generates the add and compare.

Other folds are folding it further that I don't have any part of

AZero13 replied:
The description of the test describes the input, I wasn't aware that it's meant to describe the output

AZero13 commented:
@topperc thoughts on this?

AZero13 commented:
@RKSimon Thoughts?

AZero13 committed:
[InstCombine] Pre-commit tests (NFC)

AZero13 committed:
[InstCombine] Canonicalize zext+overflow check to overflow check if zext's only purpose is to check overflow

Change processUMulZExtIdiom to also support adds, since the idiom is the same, except with add instead of mul.

Alive2: https://alive2.llvm.org/ce/z/SsB4AK
